README
ABOUT GIT-CRYPT

git-crypt enables transparent encryption and decryption of files in a
git repository.  Files which you choose to protect are encrypted when
committed, and decrypted when checked out.  git-crypt lets you freely
share a repository containing a mix of public and private content.
git-crypt gracefully degrades, so developers without the secret key can
still clone and commit to a repository with encrypted files.  This lets
you store your secret material (such as keys or passwords) in the same
repository as your code, without requiring you to lock down your entire
repository.

git-crypt was written by Andrew Ayer <agwa at andrewayer dot name>.
For more information, see <http://www.agwa.name/projects/git-crypt>.


BUILDING GIT-CRYPT

See the INSTALL file.


USING GIT-CRYPT

See the USAGE file.


CURRENT STATUS

The latest version of git-crypt is 0.3, released on 2013-04-05.
git-crypt aims to be bug-free and reliable, meaning it shouldn't
crash, malfunction, or expose your confidential data.  However,
it has not yet reached maturity, meaning it is not as documented,
featureful, or easy-to-use as it should be.  Additionally, there may be
backwards-incompatible changes introduced before version 1.0.

Development on git-crypt is currently focused on improving the user
experience, especially around setting up repositories.  There are also
plans to add additional key management schemes, such as passphrase-derived
keys and keys encrypted with PGP.


SECURITY

git-crypt is more secure that other transparent git encryption systems.
git-crypt encrypts files using AES-256 in CTR mode with a synthetic IV
derived from the SHA-1 HMAC of the file.  This is provably semantically
secure under deterministic chosen-plaintext attack.  That means that
although the encryption is deterministic (which is required so git can
distinguish when a file has and hasn't changed), it leaks no information
beyond whether two files are identical or not.  Other proposals for
transparent git encryption use ECB or CBC with a fixed IV.  These systems
are not semantically secure and leak information.

The AES key is stored unencrypted on disk.  The user is responsible for
protecting it and ensuring it's safely distributed only to authorized
people.  A future version of git-crypt may support encrypting the key
with a passphrase.


LIMITATIONS

git-crypt is not designed to encrypt an entire repository.  Not only does
that defeat the aim of git-crypt, which is the ability to selectively
encrypt files and share the repository with less-trusted developers, there
are probably better, more efficient ways to encrypt an entire repository,
such as by storing it on an encrypted filesystem.  Also note that
git-crypt is somewhat of an abuse of git's smudge, clean, and textconv
features.  Junio Hamano, git's maintainer, has said not to do this
<http://thread.gmane.org/gmane.comp.version-control.git/113124/focus=113221>,
though his main objection ("making a pair of similar 'smudged' contents
totally dissimilar in their 'clean' counterparts.") does not apply here
since git-crypt uses deterministic encryption.

git-crypt does not itself provide any authentication.  It assumes that
either the master copy of your repository is stored securely, or that
you are using git's existing facilities to ensure integrity (signed tags,
remembering commit hashes, etc.).


MAILING LISTS

To stay abreast of, and provide input to, git-crypt development, consider
subscribing to one or both of our mailing lists:

Announcements: http://lists.cloudmutt.com/mailman/listinfo/git-crypt-announce
Discussion:    http://lists.cloudmutt.com/mailman/listinfo/git-crypt-discuss


NEWS
current:
  * Updated with AGWA's recent work (v0.4)).
  * Minor improvements for compilation using mingw32.
  * Corrected error with key migration on Windows OS.
  * Updated usage documentation.
  * Updated code to compile with high warning checks.

v0.3-win32 (2013-06-17) ccleaud's changelog:
  * Ported git-crypt to Windows.
  * Added details about error when they occur.
  * Filter and diff are now set as required in .git/config

v0.3 (2013-04-05)
  * Fix 'git-crypt init' on newer versions of Git. Previously,
    encrypted files were not being automatically decrypted after
    running 'git-crypt init' with recent versions of Git.
  * Allow 'git-crypt init' to be run even if the working tree contains
    untracked files.
  * 'git-crypt init' now properly escapes arguments to the filter
    commands it configures, allowing both the path to git-crypt and the
    path to the key file to contain arbitrary characters such as spaces.

v0.2 (2013-01-25)
  * Numerous improvements to 'git-crypt init' usability.
  * Fix gitattributes example in README: the old example showed a colon
    after the filename where there shouldn't be one.
  * Various build fixes and improvements.

v0.1 (2012-11-29)
  * Initial release.


src/util.cpp
/*
 * Copyright 2012 Andrew Ayer
 * Copyright 2014 Cyril Cleaud
 *
 * This file is part of git-crypt.
 *
 * git-crypt is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * git-crypt is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with git-crypt.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Additional permission under GNU GPL version 3 section 7:
 *
 * If you modify the Program, or any covered work, by linking or
 * combining it with the OpenSSL project's OpenSSL library (or a
 * modified version of that library), containing parts covered by the
 * terms of the OpenSSL or SSLeay licenses, the licensors of the Program
 * grant you additional permission to convey the resulting work.
 * Corresponding Source for a non-source form of such a combination
 * shall include the source code for the parts of OpenSSL used as well
 * as that of the covered work.
 */

#include "util.hpp"
#include <string>
#include <cstring>
#include <cstdio>
#include <cstdlib>
#include <sys/types.h>
#if defined(__WIN32__)
#include <windows.h>
#include <stdlib.h>
#include <limits.h>
#include <io.h>
#else
#include <sys/wait.h>
#endif
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <fstream>
#include <iostream>

#if defined(__WIN32__)
char *realpath(const char *path, char resolved_path[PATH_MAX]);
bool RunRedirected(const char *p_szFileName, const char *p_szParams, int *p_piExitCode, HANDLE *p_phStdOut);
#endif

int exec_command (const char* command, std::ostream& output)
{
	int		status = 0;
#if defined(__WIN32__)
    HANDLE  hPipeOut = NULL;
    char    cmdLine[4096];
    sprintf(cmdLine, "%s%s", "cmd.exe /c ", command);
    if (RunRedirected(NULL, cmdLine, &status, &hPipeOut)) {
        DWORD dwRead=0;
        char		buffer[1024];

        while (ReadFile(hPipeOut, buffer, sizeof(buffer), &dwRead, NULL) == TRUE && dwRead > 0) {
            output.write(buffer, dwRead);
        }
        CloseHandle(hPipeOut);
    } else {
        std::clog << command << ": problem creating child process (in " << __FUNCTION__ << ")." << std::endl;
		perror("RunRedirected");
		std::exit(9);
    }
#else
	ssize_t	bytes_read;
	int		pipefd[2];
	char	buffer[1024];

	if (pipe(pipefd) == -1) {
	    std::clog << "problem creating pipe (in " << __FUNCTION__ << ")." << std::endl;
		perror("pipe");
		std::exit(9);
	}

	pid_t		child = fork();
	if (child == -1) {
	    std::clog << "problem forking (in " << __FUNCTION__ << ")." << std::endl;
		close(pipefd[0]);
		close(pipefd[1]);
		perror("fork");
		std::exit(9);
	}
	if (child == 0) {
		close(pipefd[0]);
		if (pipefd[1] != 1) {
			dup2(pipefd[1], 1);
			close(pipefd[1]);
		}
		execl("/bin/sh", "sh", "-c", command, NULL);
		exit(-1);
	}
	close(pipefd[1]);

	while ((bytes_read = read(pipefd[0], buffer, sizeof(buffer))) > 0) {
		output.write(buffer, bytes_read);
	}
	close(pipefd[0]);
	waitpid(child, &status, 0);
#endif
	return status;
}

std::string resolve_path (const char* path)
{
	char*		resolved_path_p = realpath(path, NULL);
	std::string	resolved_path(resolved_path_p);
	free(resolved_path_p);
	return resolved_path;
}

void	open_tempfile (std::fstream& file, std::ios_base::openmode mode)
{
#if defined(__WIN32__)
	const char*	tmpdir = getenv("TEMP");
#else
	const char*	tmpdir = getenv("TMPDIR");
#endif
	size_t		tmpdir_len;

	if (tmpdir) {
		tmpdir_len = strlen(tmpdir);
	} else {
		tmpdir = "/tmp";
		tmpdir_len = 4;
	}
	char*		path = new char[tmpdir_len + 18];
	strcpy(path, tmpdir);
	strcpy(path + tmpdir_len, "/git-crypt.XXXXXX");

#if defined(__WIN32__)
	char *result = _mktemp(path);
	if(result != NULL) {
		path = result;
	} else {
        std::clog << "problem with _mktemp (in " << __FUNCTION__ << ")." << std::endl;
		perror("_mktemp");
		std::exit(9);
	}
#else
	mode_t		old_umask = umask(0077);
	int		fd = mkstemp(path);
	if (fd == -1) {
	    std::clog << "problem with mkstemp (in " << __FUNCTION__ << ")." << std::endl;
		perror("mkstemp");
		std::exit(9);
	}
	umask(old_umask);
#endif

	file.open(path, mode);
	if (!file.is_open()) {
	    std::clog << path << ": problem opening file (in " << __FUNCTION__ << ")." << std::endl;
		perror("open");
		unlink(path);
		std::exit(9);
	}
	unlink(path);

#if !defined(__WIN32__)
	close(fd);
#endif
	delete[] path;
}

std::string	escape_shell_arg (const std::string& str)
{
	std::string	new_str;
	new_str.push_back('"');
	for (std::string::const_iterator it(str.begin()); it != str.end(); ++it) {
		if (*it == '"' || *it == '\\' || *it == '$' || *it == '`') {
			new_str.push_back('\\');
		}
		new_str.push_back(*it);
	}
	new_str.push_back('"');
	return new_str;
}


#if defined(__WIN32__)
char *realpath(const char *path, char resolved_path[PATH_MAX])
{
    char *return_path = 0;

    if (path) { //Else EINVAL
        if (resolved_path)
            return_path = resolved_path;
        else        //Non standard extension that glibc uses
            return_path = (char *)malloc(PATH_MAX);

        if (return_path) { //Else EINVAL
            //This is a Win32 API function similar to what realpath() is supposed to do
            size_t size = GetFullPathNameA(path, PATH_MAX, return_path, 0);

            //GetFullPathNameA() returns a size larger than buffer if buffer is too small
            if (size > PATH_MAX) {
                if (return_path != resolved_path) { //Malloc'd buffer - Unstandard extension retry
                    free(return_path);
                    return_path = (char *)malloc(size);

                    if (return_path) {
                        size_t new_size = GetFullPathNameA(path, size, return_path, 0); //Try again
                        if (new_size > size) { //If it's still too large, we have a problem, don't try again
                            free(return_path);
                            return_path = 0;
                            errno = ENAMETOOLONG;
                        } else {
                            size = new_size;
                        }
                    } else {
                        //I wasn't sure what to return here, but the standard does say to return EINVAL
                        //if resolved_path is null, and in this case we couldn't malloc large enough buffer
                        errno = EINVAL;
                    }
                } else { //resolved_path buffer isn't big enough
                    return_path = 0;
                    errno = ENAMETOOLONG;
                }
            }

            //GetFullPathNameA() returns 0 if some path resolve problem occured
            if (!size) {
                if (return_path != resolved_path) //Malloc'd buffer
                    free(return_path);

                return_path = 0;

                //Convert MS errors into standard errors
                switch (GetLastError()) {
                case ERROR_FILE_NOT_FOUND:
                    errno = ENOENT;
                    break;

                case ERROR_PATH_NOT_FOUND:
                case ERROR_INVALID_DRIVE:
                    errno = ENOTDIR;
                    break;

                case ERROR_ACCESS_DENIED:
                    errno = EACCES;
                    break;

                default: //Unknown Error
                    errno = EIO;
                    break;
                }
            }

            //If we get to here with a valid return_path, we're still doing good
            if (return_path) {
                struct stat stat_buffer;

                //Make sure path exists, stat() returns 0 on success
                if (stat(return_path, &stat_buffer)) {
                    if (return_path != resolved_path)
                        free(return_path);

                    return_path = 0;
                    //stat() will set the correct errno for us
                }
                //else we succeeded!
            }
        }
        else
            errno = EINVAL;
    }
    else
        errno = EINVAL;

    return return_path;
}

bool RunRedirected(const char *p_szFileName, const char *p_szParams, int *p_piExitCode, HANDLE *p_phStdOut)
{
    bool bResult=false;
    HANDLE hInput=NULL, hOutput=NULL;
    PROCESS_INFORMATION structProcessInfo= {0, 0, 0, 0};
    STARTUPINFO structProcessStartup;
    SECURITY_ATTRIBUTES saAttr;


    // Buffers initialization
    memset(&structProcessStartup, 0, sizeof(structProcessStartup));
    memset(&saAttr, 0, sizeof(saAttr));

    // Prepare process creation
    structProcessStartup.cb=sizeof(structProcessStartup);
    structProcessStartup.dwX=100;
    structProcessStartup.dwY=100;
    structProcessStartup.dwFlags=STARTF_USESHOWWINDOW|STARTF_USEPOSITION|STARTF_USESTDHANDLES;
    structProcessStartup.wShowWindow=SW_HIDE;

    // Default redirections
    structProcessStartup.hStdOutput=GetStdHandle(STD_OUTPUT_HANDLE);

    // Set the bInheritHandle flag so pipe handles are inherited.
    saAttr.nLength=sizeof(SECURITY_ATTRIBUTES);
    saAttr.bInheritHandle=TRUE;
    saAttr.lpSecurityDescriptor=NULL;

    // Create a pipe for the child process's STDOUT.
    if(CreatePipe(&hOutput, &hInput, &saAttr, 512 * 1024) != FALSE) {
        if(p_phStdOut)
            *p_phStdOut=hOutput;
        structProcessStartup.hStdOutput=hInput;

        if(CreateProcess(NULL, (LPSTR)p_szParams, NULL, NULL, TRUE,
                         CREATE_DEFAULT_ERROR_MODE, NULL, NULL, &structProcessStartup, &structProcessInfo) != FALSE) {
            // Wait for process to end
            WaitForSingleObject(structProcessInfo.hProcess, INFINITE);

            // Get exit code if requested by caller
            if(p_piExitCode)
                GetExitCodeProcess(structProcessInfo.hProcess, (PDWORD)p_piExitCode);

            // Close unneeded handle if needed
            CloseHandle(structProcessStartup.hStdOutput);
            CloseHandle(structProcessInfo.hProcess);
            CloseHandle(structProcessInfo.hThread);
            bResult = true;
        }
    }

    return bResult;
}
#endif


src/util.hpp
/*
 * Copyright 2012, 2014 Andrew Ayer
 *
 * This file is part of git-crypt.
 *
 * git-crypt is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * git-crypt is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with git-crypt.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Additional permission under GNU GPL version 3 section 7:
 *
 * If you modify the Program, or any covered work, by linking or
 * combining it with the OpenSSL project's OpenSSL library (or a
 * modified version of that library), containing parts covered by the
 * terms of the OpenSSL or SSLeay licenses, the licensors of the Program
 * grant you additional permission to convey the resulting work.
 * Corresponding Source for a non-source form of such a combination
 * shall include the source code for the parts of OpenSSL used as well
 * as that of the covered work.
 */

#ifndef GIT_CRYPT_UTIL_HPP
#define GIT_CRYPT_UTIL_HPP

#include <string>
#include <ios>
#include <iosfwd>
#include <stdint.h>
#include <fstream>
#include <vector>

struct System_error {
	std::string	action;
	std::string	target;
	int		error;

	System_error (const std::string& a, const std::string& t, int e) : action(a), target(t), error(e) { }

	std::string message () const;
};

class temp_fstream : public std::fstream {
	std::string	filename;
public:
	~temp_fstream () { close(); }

	void		open (std::ios_base::openmode);
	void		close ();
};

void		mkdir_parent (const std::string& path); // Create parent directories of path, __but not path itself__
std::string	our_exe_path ();
int			exec_command (const std::vector<std::string>&);
int			exec_command (const std::vector<std::string>&, std::ostream& output);
int			exec_command_with_input (const std::vector<std::string>&, const char* p, size_t len);
bool		successful_exit (int status);
std::string	escape_shell_arg (const std::string&);
uint32_t	load_be32 (const unsigned char*);
void		store_be32 (unsigned char*, uint32_t);
bool		read_be32 (std::istream& in, uint32_t&);
void		write_be32 (std::ostream& out, uint32_t);
void		init_std_streams ();

mode_t 		util_umask (mode_t);
int			util_rename (const char*, const char*);

#ifndef UNREFERENCED_PARAMETER
#define UNREFERENCED_PARAMETER(P) {(P)=(P);}
#endif

#endif



src/crypto.cpp
/*
 * Copyright 2012 Andrew Ayer
 *
 * This file is part of git-crypt.
 *
 * git-crypt is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * git-crypt is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with git-crypt.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Additional permission under GNU GPL version 3 section 7:
 *
 * If you modify the Program, or any covered work, by linking or
 * combining it with the OpenSSL project's OpenSSL library (or a
 * modified version of that library), containing parts covered by the
 * terms of the OpenSSL or SSLeay licenses, the licensors of the Program
 * grant you additional permission to convey the resulting work.
 * Corresponding Source for a non-source form of such a combination
 * shall include the source code for the parts of OpenSSL used as well
 * as that of the covered work.
 */

#define _BSD_SOURCE
#include "crypto.hpp"
#include <openssl/aes.h>
#include <openssl/sha.h>
#include <openssl/hmac.h>
#include <openssl/evp.h>
#include <fstream>
#include <iostream>
#include <cstring>
#include <cstdlib>
#if defined(__WIN32__)
#include <winsock2.h>
#else
#include <arpa/inet.h>
#endif

void load_keys (const char* filepath, keys_t* keys)
{
	std::ifstream	file(filepath, std::ios::binary);
	if (!file) {
		std::clog << filepath << ": problem loading file (in " << __FUNCTION__ << ")." << std::endl;
		perror(filepath);
		std::exit(1);
	}
	char	buffer[AES_KEY_BITS/8 + HMAC_KEY_LEN];
	file.read(buffer, sizeof(buffer));
	if (file.gcount() != sizeof(buffer)) {
		std::clog << filepath << ": premature end of key file\n";
		std::exit(1);
	}

	// First comes the AES encryption key
	if (AES_set_encrypt_key(reinterpret_cast<uint8_t*>(buffer), AES_KEY_BITS, &keys->enc) != 0) {
		std::clog << filepath << ": failed to initialize AES encryption key\n";
		std::exit(1);
	}

	// Then it's the HMAC key
	memcpy(keys->hmac, buffer + AES_KEY_BITS/8, HMAC_KEY_LEN);
}


aes_ctr_state::aes_ctr_state (const uint8_t* arg_nonce, size_t arg_nonce_len)
{
	memset(nonce, '\0', sizeof(nonce));
	memcpy(nonce, arg_nonce, std::min(arg_nonce_len, sizeof(nonce)));
	byte_counter = 0;
	memset(otp, '\0', sizeof(otp));
}

void aes_ctr_state::process (const AES_KEY* key, const uint8_t* in, uint8_t* out, size_t len)
{
	for (size_t i = 0; i < len; ++i) {
		if (byte_counter % 16 == 0) {
			// Generate a new OTP
			// CTR value:
			//  first 12 bytes - nonce
			//  last   4 bytes - block number (sequentially increasing with each block)
			uint8_t		ctr[16];
			uint32_t	blockno = htonl(byte_counter / 16);
			memcpy(ctr, nonce, 12);
			memcpy(ctr + 12, &blockno, 4);
			AES_encrypt(ctr, otp, key);
		}

		// encrypt one byte
		out[i] = in[i] ^ otp[byte_counter++ % 16];
	}
}

hmac_sha1_state::hmac_sha1_state (const uint8_t* key, size_t key_len)
{
	HMAC_Init(&ctx, key, key_len, EVP_sha1());
}

hmac_sha1_state::~hmac_sha1_state ()
{
	HMAC_cleanup(&ctx);
}

void hmac_sha1_state::add (const uint8_t* buffer, size_t buffer_len)
{
	HMAC_Update(&ctx, buffer, buffer_len);
}

void hmac_sha1_state::get (uint8_t* digest)
{
	unsigned int len;
	HMAC_Final(&ctx, digest, &len);
}


// Encrypt/decrypt an entire input stream, writing to the given output stream
void process_stream (std::istream& in, std::ostream& out, const AES_KEY* enc_key, const uint8_t* nonce)
{
	aes_ctr_state	state(nonce, 12);

	uint8_t		buffer[1024];
	while (in) {
		in.read(reinterpret_cast<char*>(buffer), sizeof(buffer));
		state.process(enc_key, buffer, buffer, in.gcount());
		out.write(reinterpret_cast<char*>(buffer), in.gcount());
	}
}


src/util-win32.cpp
/*
 * Copyright 2014 Andrew Ayer
 *
 * This file is part of git-crypt.
 *
 * git-crypt is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * git-crypt is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with git-crypt.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Additional permission under GNU GPL version 3 section 7:
 *
 * If you modify the Program, or any covered work, by linking or
 * combining it with the OpenSSL project's OpenSSL library (or a
 * modified version of that library), containing parts covered by the
 * terms of the OpenSSL or SSLeay licenses, the licensors of the Program
 * grant you additional permission to convey the resulting work.
 * Corresponding Source for a non-source form of such a combination
 * shall include the source code for the parts of OpenSSL used as well
 * as that of the covered work.
 */

#include <io.h>
#include <stdio.h>
#include <fcntl.h>
#include <windows.h>
#include <vector>

std::string System_error::message () const
{
	std::string	mesg(action);
	if (!target.empty()) {
		mesg += ": ";
		mesg += target;
	}
	if (error) {
		LPTSTR	error_message;
		FormatMessageA(
			FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
			NULL,
			error,
			MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
			reinterpret_cast<LPTSTR>(&error_message),
			0,
			NULL);
		mesg += error_message;
		LocalFree(error_message);
	}
	return mesg;
}

void	temp_fstream::open (std::ios_base::openmode mode)
{
	close();

	char			tmpdir[MAX_PATH + 1];

	DWORD			ret = GetTempPath(sizeof(tmpdir), tmpdir);
	if (ret == 0) {
		throw System_error("GetTempPath", "", GetLastError());
	} else if (ret > sizeof(tmpdir) - 1) {
		throw System_error("GetTempPath", "", ERROR_BUFFER_OVERFLOW);
	}

	char			tmpfilename[MAX_PATH + 1];
	if (GetTempFileName(tmpdir, TEXT("git-crypt"), 0, tmpfilename) == 0) {
		throw System_error("GetTempFileName", "", GetLastError());
	}

	filename = tmpfilename;

	std::fstream::open(filename.c_str(), mode);
	if (!std::fstream::is_open()) {
		DeleteFile(filename.c_str());
		throw System_error("std::fstream::open", filename, 0);
	}
}

void	temp_fstream::close ()
{
	if (std::fstream::is_open()) {
		std::fstream::close();
		DeleteFile(filename.c_str());
	}
}

void	mkdir_parent (const std::string& path)
{
	std::string::size_type		slash(path.find('/', 1));
	while (slash != std::string::npos) {
		std::string		prefix(path.substr(0, slash));
		if (GetFileAttributes(prefix.c_str()) == INVALID_FILE_ATTRIBUTES) {
			// prefix does not exist, so try to create it
			if (!CreateDirectory(prefix.c_str(), NULL)) {
				throw System_error("CreateDirectory", prefix, GetLastError());
			}
		}

		slash = path.find('/', slash + 1);
	}
}

std::string our_exe_path ()
{
	std::vector<char>	buffer(128);
	size_t			len;

	while ((len = GetModuleFileNameA(NULL, &buffer[0], buffer.size())) == buffer.size()) {
		// buffer may have been truncated - grow and try again
		buffer.resize(buffer.size() * 2);
	}
	if (len == 0) {
		throw System_error("GetModuleFileNameA", "", GetLastError());
	}

	return std::string(buffer.begin(), buffer.begin() + len);
}

static void escape_cmdline_argument (std::string& cmdline, const std::string& arg)
{
	// For an explanation of Win32's arcane argument quoting rules, see:
	//  http://msdn.microsoft.com/en-us/library/17w5ykft%28v=vs.85%29.aspx
	//  http://msdn.microsoft.com/en-us/library/bb776391%28v=vs.85%29.aspx
	//  http://blogs.msdn.com/b/twistylittlepassagesallalike/archive/2011/04/23/everyone-quotes-arguments-the-wrong-way.aspx
	//  http://blogs.msdn.com/b/oldnewthing/archive/2010/09/17/10063629.aspx
	cmdline.push_back('"');

	std::string::const_iterator	p(arg.begin());
	while (p != arg.end()) {
		if (*p == '"') {
			cmdline.push_back('\\');
			cmdline.push_back('"');
			++p;
		} else if (*p == '\\') {
			unsigned int	num_backslashes = 0;
			while (p != arg.end() && *p == '\\') {
				++num_backslashes;
				++p;
			}
			if (p == arg.end() || *p == '"') {
				// Backslashes need to be escaped
				num_backslashes *= 2;
			}
			while (num_backslashes--) {
				cmdline.push_back('\\');
			}
		} else {
			cmdline.push_back(*p++);
		}
	}

	cmdline.push_back('"');
}

static std::string format_cmdline (const std::vector<std::string>& command)
{
	std::string		cmdline;
	for (std::vector<std::string>::const_iterator arg(command.begin()); arg != command.end(); ++arg) {
		if (arg != command.begin()) {
			cmdline.push_back(' ');
		}
		escape_cmdline_argument(cmdline, *arg);
	}
	return cmdline;
}

static int wait_for_child (HANDLE child_handle)
{
	if (WaitForSingleObject(child_handle, INFINITE) == WAIT_FAILED) {
		throw System_error("WaitForSingleObject", "", GetLastError());
	}

	DWORD			exit_code;
	if (!GetExitCodeProcess(child_handle, &exit_code)) {
		throw System_error("GetExitCodeProcess", "", GetLastError());
	}

	return exit_code;
}

static HANDLE spawn_command (const std::vector<std::string>& command, HANDLE stdin_handle, HANDLE stdout_handle, HANDLE stderr_handle)
{
	PROCESS_INFORMATION	proc_info;
	ZeroMemory(&proc_info, sizeof(proc_info));

	STARTUPINFO		start_info;
	ZeroMemory(&start_info, sizeof(start_info));

	start_info.cb = sizeof(STARTUPINFO);
	start_info.hStdInput = stdin_handle ? stdin_handle : GetStdHandle(STD_INPUT_HANDLE);
	start_info.hStdOutput = stdout_handle ? stdout_handle : GetStdHandle(STD_OUTPUT_HANDLE);
	start_info.hStdError = stderr_handle ? stderr_handle : GetStdHandle(STD_ERROR_HANDLE);
	start_info.dwFlags |= STARTF_USESTDHANDLES;

	std::string		cmdline(format_cmdline(command));

	if (!CreateProcessA(NULL,		// application name (NULL to use command line)
				const_cast<char*>(cmdline.c_str()),
				NULL,		// process security attributes
				NULL,		// primary thread security attributes
				TRUE,		// handles are inherited
				0,		// creation flags
				NULL,		// use parent's environment
				NULL,		// use parent's current directory
				&start_info,
				&proc_info)) {
		throw System_error("CreateProcess", cmdline, GetLastError());
	}

	CloseHandle(proc_info.hThread);

	return proc_info.hProcess;
}

int exec_command (const std::vector<std::string>& command)
{
	HANDLE			child_handle = spawn_command(command, NULL, NULL, NULL);
	int			exit_code = wait_for_child(child_handle);
	CloseHandle(child_handle);
	return exit_code;
}

int exec_command (const std::vector<std::string>& command, std::ostream& output)
{
	HANDLE			stdout_pipe_reader = NULL;
	HANDLE			stdout_pipe_writer = NULL;
	SECURITY_ATTRIBUTES	sec_attr;

	// Set the bInheritHandle flag so pipe handles are inherited.
	sec_attr.nLength = sizeof(SECURITY_ATTRIBUTES);
	sec_attr.bInheritHandle = TRUE;
	sec_attr.lpSecurityDescriptor = NULL;

	// Create a pipe for the child process's STDOUT.
	if (!CreatePipe(&stdout_pipe_reader, &stdout_pipe_writer, &sec_attr, 0)) {
		throw System_error("CreatePipe", "", GetLastError());
	}

	// Ensure the read handle to the pipe for STDOUT is not inherited.
	if (!SetHandleInformation(stdout_pipe_reader, HANDLE_FLAG_INHERIT, 0)) {
		throw System_error("SetHandleInformation", "", GetLastError());
	}

	HANDLE			child_handle = spawn_command(command, NULL, stdout_pipe_writer, NULL);
	CloseHandle(stdout_pipe_writer);

	// Read from stdout_pipe_reader.
	// Note that ReadFile on a pipe may return with bytes_read==0 if the other
	// end of the pipe writes zero bytes, so don't break out of the read loop
	// when this happens.  When the other end of the pipe closes, ReadFile
	// fails with ERROR_BROKEN_PIPE.
	char			buffer[1024];
	DWORD			bytes_read;
	while (ReadFile(stdout_pipe_reader, buffer, sizeof(buffer), &bytes_read, NULL)) {
		output.write(buffer, bytes_read);
	}
	const DWORD		read_error = GetLastError();
	if (read_error != ERROR_BROKEN_PIPE) {
		throw System_error("ReadFile", "", read_error);
	}

	CloseHandle(stdout_pipe_reader);

	int			exit_code = wait_for_child(child_handle);
	CloseHandle(child_handle);
	return exit_code;
}

int exec_command_with_input (const std::vector<std::string>& command, const char* p, size_t len)
{
	HANDLE			stdin_pipe_reader = NULL;
	HANDLE			stdin_pipe_writer = NULL;
	SECURITY_ATTRIBUTES	sec_attr;

	// Set the bInheritHandle flag so pipe handles are inherited.
	sec_attr.nLength = sizeof(SECURITY_ATTRIBUTES);
	sec_attr.bInheritHandle = TRUE;
	sec_attr.lpSecurityDescriptor = NULL;

	// Create a pipe for the child process's STDIN.
	if (!CreatePipe(&stdin_pipe_reader, &stdin_pipe_writer, &sec_attr, 0)) {
		throw System_error("CreatePipe", "", GetLastError());
	}

	// Ensure the write handle to the pipe for STDIN is not inherited.
	if (!SetHandleInformation(stdin_pipe_writer, HANDLE_FLAG_INHERIT, 0)) {
		throw System_error("SetHandleInformation", "", GetLastError());
	}

	HANDLE			child_handle = spawn_command(command, stdin_pipe_reader, NULL, NULL);
	CloseHandle(stdin_pipe_reader);

	// Write to stdin_pipe_writer.
	while (len > 0) {
		DWORD		bytes_written;
		if (!WriteFile(stdin_pipe_writer, p, len, &bytes_written, NULL)) {
			throw System_error("WriteFile", "", GetLastError());
		}
		p += bytes_written;
		len -= bytes_written;
	}

	CloseHandle(stdin_pipe_writer);

	int			exit_code = wait_for_child(child_handle);
	CloseHandle(child_handle);
	return exit_code;
}

bool successful_exit (int status)
{
	return status == 0;
}

static void	init_std_streams_platform ()
{
	_setmode(_fileno(stdin), _O_BINARY);
	_setmode(_fileno(stdout), _O_BINARY);
}

mode_t util_umask (mode_t fmask)
{
	UNREFERENCED_PARAMETER(fmask);
	// Not available in Windows and function not always defined in Win32 environments
	return 0;
}

int util_rename (const char* from, const char* to)
{
	// On Windows OS, it is necessary to ensure target file doesn't exist
	unlink(to);
	return rename(from, to);
}
